File: ios/Godstone/Sources/GodstoneMesh/PanicWipe.swift
import Foundation
import Security

// MARK: - Coordinated, resumable, cryptographic-erasure panic wipe

/// The persisted phase of an in-progress wipe (ADR-004 criterion 5,
/// GST-WIPE-001). Ordered by the `PanicWipe.run()` ladder.
public enum WipeState: String, Sendable {
    case idle
    case requested
    case keyErased
    case artifactsDeleted
    case newIdentity
}

/// Crash-safe marker for the wipe state machine. Implementations must persist
/// across a reboot/jetsam so `PanicWipe.resumeIfPending()` can finish an
/// interrupted wipe.
public protocol WipeJournal: AnyObject {
    func read() -> WipeState
    func write(_ state: WipeState)
    func clear()
}

/// The three idempotent destroy/rebuild steps, injectable so the state machine
/// is host-testable without the Keychain. See `PanicWipe` for the ordering
/// contract.
public protocol WipeArtifacts: AnyObject {
    /// Destroy the long-lived secret keys. After this, prior data is
    /// cryptographically erased (unrecoverable) even if its container still
    /// exists. This is the point of no return. Throwing propagates as a wipe
    /// failure: the journal is NOT advanced past this step, so the next
    /// `resumeIfPending()` retries it.
    func eraseKeys() throws
    /// Delete the now-useless artifact containers (ciphertext files, DBs,
    /// prefs). Idempotent. On iOS this deletes the durable DTN store DB file
    /// (Phase G); the Keychain keys themselves are removed by `eraseKeys`.
    func deleteArtifacts() throws
    /// Generate + store a fresh identity.
    func regenerateIdentity() throws
}

/// A small persisted state machine that destroys long-lived secrets across the
/// store + identity + the key material that protects them, then regenerates a
/// fresh identity so prior traffic cannot be linked to the new node.
///
/// It is crash-safe: the state is written to a durable `WipeJournal` AFTER each
/// step completes, and every step is idempotent, so a crash-then-resume re-runs
/// at most the one step that was interrupted and then continues forward. The
/// ordering is deliberate:
///
///     idle -> requested -> keyErased -> artifactsDeleted -> newIdentity -> idle
///
///       requested        -- wipe requested, nothing destroyed yet
///       eraseKeys()      -- destroy the keys. After this the data is
///                           cryptographically erased. Point of no return.
///       keyErased        -- keys gone, containers may still exist but unreadable
///       deleteArtifacts()-- delete the now-useless containers. Idempotent.
///       artifactsDeleted -- no recoverable artifacts remain
///       regenerateIdentity() -- create a fresh identity.
///       newIdentity      -- new identity in place
///       clear()          -- drop the journal; back to idle.
///
/// Crypto-erasure-first is the safety property: even a total failure after
/// `eraseKeys()` cannot leave prior data recoverable. The remaining steps are
/// cleanup + re-identity, and the journal guarantees they eventually run.
///
/// The machine itself is pure (no Keychain/Security/UserDefaults touch); the
/// platform glue (`UserDefaultsWipeJournal`, `KeychainWipeArtifacts`) is
/// injected. `PanicWipeTests` drives the machine with fakes -- simulating a
/// crash before each step -- and asserts resumability + ordering + the
/// no-op-when-idle contract without a device or Keychain.
public final class PanicWipe {
    private let journal: WipeJournal
    private let artifacts: WipeArtifacts

    public init(journal: WipeJournal, artifacts: WipeArtifacts) {
        self.journal = journal
        self.artifacts = artifacts
    }

    /// Begin a wipe from a clean (idle) state. Writes `requested`, then
    /// advances as far as it can. A step that throws (a wipe failure / simulated
    /// crash) propagates: the journal is left at the last completed step, so the
    /// next `resumeIfPending()` finishes it.
    public func begin() throws {
        journal.write(.requested)
        try run()
    }

    /// If a wipe was in progress (journal != idle), finish it. Safe to call on
    /// every app launch. No-op when no wipe is pending.
    public func resumeIfPending() throws {
        guard journal.read() != .idle else { return }
        try run()
    }

    /// Advance from the persisted state to completion. Each rung checks the
    /// persisted state, performs that rung's step, then persists the next state
    /// -- so a throw between perform and persist re-runs that one idempotent
    /// step on resume, and a throw after persist skips it. The cascade runs
    /// every remaining rung unless a step throws; the journal then reflects the
    /// last completed rung and the next resume continues from there.
    private func run() throws {
        var s = journal.read()
        if s == .idle { s = .requested; journal.write(s) }
        if s == .requested {
            try artifacts.eraseKeys()
            s = .keyErased; journal.write(s)
        }
        if s == .keyErased {
            try artifacts.deleteArtifacts()
            s = .artifactsDeleted; journal.write(s)
        }
        if s == .artifactsDeleted {
            try artifacts.regenerateIdentity()
            s = .newIdentity; journal.write(s)
        }
        if s == .newIdentity {
            journal.clear()
        }
    }

    // MARK: - Production entry points

    /// Production entry point: start a full wipe using platform glue.
    public static func begin(journal: WipeJournal = UserDefaultsWipeJournal(),
                             artifacts: WipeArtifacts = KeychainWipeArtifacts()) throws {
        try PanicWipe(journal: journal, artifacts: artifacts).begin()
    }

    /// Call on app launch: finishes a wipe that a crash interrupted.
    public static func resumeIfPending(journal: WipeJournal = UserDefaultsWipeJournal(),
                                       artifacts: WipeArtifacts = KeychainWipeArtifacts()) throws {
        try PanicWipe(journal: journal, artifacts: artifacts).resumeIfPending()
    }
}

// MARK: - Production glue

/// Journal backed by UserDefaults. The marker is a single state string; it is
/// not sensitive and a leftover `.idle` marker is harmless. A non-idle marker
/// after a reboot is exactly the signal `resumeIfPending` acts on.
public final class UserDefaultsWipeJournal: WipeJournal {
    private let key = "io.godstone.wipe.state"
    private let defaults: UserDefaults

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    public func read() -> WipeState {
        guard let raw = defaults.string(forKey: key) else { return .idle }
        return WipeState(rawValue: raw) ?? .idle
    }

    public func write(_ state: WipeState) {
        defaults.set(state.rawValue, forKey: key)
    }

    public func clear() {
        defaults.removeObject(forKey: key)
    }
}

/// Production `WipeArtifacts` for iOS.
///
/// `eraseKeys` deletes the two Keychain identity items via
/// `MeshIdentity.deleteFromKeychain()` -- on iOS the private keys ARE the
/// secret (there is no separate KEK wrapping ciphertext files as on Android),
/// so this is both key destruction and artifact deletion in one step.
/// `deleteArtifacts` deletes the durable DTN store DB file (Phase G) when a
/// `storeUrl` is registered; the Keychain keys are already gone by this point
/// (eraseKeys), so this is cleanup of now-useless artifacts, coordinated with
/// the store + identity wipe as on Android. `regenerateIdentity` builds a fresh
/// identity via `MeshIdentity.generateAndStore()`.
///
/// Each step is idempotent: `SecItemDelete` on an absent item is a no-op,
/// `SqliteMessageStore.panicWipe(at:)` on an absent file is a no-op, and
/// `generateAndStore` creates fresh keys (deleting any same-tag item first).
public final class KeychainWipeArtifacts: WipeArtifacts {
    /// URL of the durable store DB file to wipe in `deleteArtifacts`. Nil when
    /// no durable store is configured (the wipe then only touches the Keychain).
    private let storeUrl: URL?

    /// `storeUrl` registers the durable DTN store for coordinated wipe.
    public init(storeUrl: URL? = nil) {
        self.storeUrl = storeUrl
    }

    public func eraseKeys() throws {
        // Crypto erasure: the keys are the secret. Deleting them makes prior
        // traffic permanently unlinkable. Idempotent (errSecItemNotFound ok).
        try MeshIdentity.deleteFromKeychain()
    }

    public func deleteArtifacts() {
        // Delete the durable store DB file (+ WAL/SHM). The Keychain keys are
        // already gone (eraseKeys); this is cleanup of the now-useless artifact
        // container, coordinated with the identity wipe. Idempotent.
        if let url = storeUrl {
            SqliteMessageStore.panicWipe(at: url)
        }
    }

    public func regenerateIdentity() throws {
        _ = try MeshIdentity.generateAndStore()
    }
}

File: android/mesh/src/main/java/io/godstone/mesh/identity/PanicWipe.kt
package io.godstone.mesh.identity

import android.content.Context
import io.godstone.mesh.store.SqliteMessageStore
import java.security.KeyStore

/**
 * Coordinated, resumable, cryptographic-erasure panic wipe (ADR-004 criterion 5,
 * GST-WIPE-001).
 *
 * The wipe destroys long-lived secrets across the store + identity + the key
 * material that encrypts them, then regenerates a fresh identity so prior
 * traffic cannot be linked to the new node. It must be crash-safe: a process
 * kill, power loss or jetsam at any point must leave the device in a state from
 * which a restart finishes the wipe without leaking recoverable data and without
 * skipping a step.
 *
 * The construction is a small persisted state machine. The state is written to a
 * durable [WipeJournal] AFTER each step completes, and every step is idempotent,
 * so a crash-then-resume re-runs at most the one step that was interrupted and
 * then continues forward. The ordering is deliberate:
 *
 *   IDLE -> REQUESTED -> KEY_ERASED -> ARTIFACTS_DELETED -> NEW_IDENTITY -> IDLE
 *
 *   REQUESTED          -- wipe requested, nothing destroyed yet
 *   eraseKeys()        -- destroy the KEK (Android Keystore master key). After
 *                         this the data is cryptographically erased: the
 *                         EncryptedSharedPreferences ciphertexts for the identity
 *                         and the store key are undecryptable even if their
 *                         files still sit on disk. KEY_ERASED is the point of no
 *                         return.
 *   KEY_ERASED         -- KEK gone, ciphertext files still present but unreadable
 *   deleteArtifacts()  -- delete the now-useless ciphertext files (store DB, store
 *                         key prefs, identity prefs). Idempotent.
 *   ARTIFACTS_DELETED  -- no recoverable artifacts remain
 *   regenerateIdentity()-- create a fresh Keystore master key + fresh identity.
 *   NEW_IDENTITY       -- new identity in place
 *   clear()            -- drop the journal; the device is back to IDLE clean.
 *
 * Crypto-erasure-first is the safety property: even a total failure after
 * [WipeArtifacts.eraseKeys] cannot leave prior data recoverable, because the key
 * that decrypted it is gone. The remaining steps are cleanup + re-identity, and
 * the journal guarantees they eventually run.
 *
 * The state machine itself ([PanicWipe]) is pure Kotlin with no Android
 * dependency; the platform glue ([FileWipeJournal], [AndroidWipeArtifacts]) is
 * injected. The host unit tests in `PanicWipeTest` drive the machine with fakes
 * (simulating a crash before each step) and assert resumability + ordering +
 * the no-op-when-idle contract without a device. The production glue is thin
 * (Keystore entry delete + the existing `SqliteMessageStore.panicWipe` /
 * `Identity.panicWipe` / `Identity.loadOrCreate`); the actual Keystore + file
 * deletion is an on-device verification, not host-testable.
 */
class PanicWipe(
    private val journal: WipeJournal,
    private val artifacts: WipeArtifacts,
) {

    /** The persisted phase of an in-progress wipe. Ordered by the run() ladder. */
    enum class WipeState {
        IDLE,
        REQUESTED,
        KEY_ERASED,
        ARTIFACTS_DELETED,
        NEW_IDENTITY,
    }

    /**
     * Begin a wipe from a clean (IDLE) state. Writes REQUESTED, then advances as
     * far as it can. A crash at any point leaves the journal at the last
     * completed step; [resumeIfPending] on the next launch finishes it.
     */
    fun begin() {
        journal.write(WipeState.REQUESTED)
        run()
    }

    /**
     * If a wipe was in progress (journal != IDLE), finish it. Safe to call on
     * every app launch. No-op when no wipe is pending.
     */
    fun resumeIfPending() {
        if (journal.read() != WipeState.IDLE) run()
    }

    /**
     * Advance the machine from its persisted state to completion. Each rung
     * checks the persisted state, performs the step that rung owns, then
     * persists the next state -- so a crash between "perform" and "persist"
     * re-runs that one idempotent step on resume, and a crash after "persist"
     * skips it. The cascade falls through: a single call runs every remaining
     * rung unless a step throws (simulating a crash), in which case the journal
     * already reflects the last completed rung and the next resume continues.
     */
    private fun run() {
        var s = journal.read()
        if (s == WipeState.IDLE) { s = WipeState.REQUESTED; journal.write(s) }
        if (s == WipeState.REQUESTED) {
            artifacts.eraseKeys()
            s = WipeState.KEY_ERASED; journal.write(s)
        }
        if (s == WipeState.KEY_ERASED) {
            artifacts.deleteArtifacts()
            s = WipeState.ARTIFACTS_DELETED; journal.write(s)
        }
        if (s == WipeState.ARTIFACTS_DELETED) {
            artifacts.regenerateIdentity()
            s = WipeState.NEW_IDENTITY; journal.write(s)
        }
        if (s == WipeState.NEW_IDENTITY) {
            journal.clear()
        }
    }

    companion object {
        /** Production entry point: start a full wipe using platform glue. */
        fun begin(ctx: Context) {
            PanicWipe(FileWipeJournal(ctx), AndroidWipeArtifacts(ctx)).begin()
        }

        /** Call on app launch: finishes a wipe that a crash interrupted. */
        fun resumeIfPending(ctx: Context) {
            PanicWipe(FileWipeJournal(ctx), AndroidWipeArtifacts(ctx)).resumeIfPending()
        }
    }
}

/** Crash-safe marker for the wipe state machine. Implementations must persist. */
interface WipeJournal {
    fun read(): PanicWipe.WipeState
    fun write(state: PanicWipe.WipeState)
    fun clear()
}

/** The three idempotent destroy/rebuild steps, injectable for host tests. */
interface WipeArtifacts {
    /** Destroy the KEK. After this, encrypted artifacts are unrecoverable. */
    fun eraseKeys()
    /** Delete the ciphertext/DB/prefs files. Idempotent. */
    fun deleteArtifacts()
    /** Generate + store a fresh identity (and the KEK that protects it). */
    fun regenerateIdentity()
}

/**
 * Journal backed by a dedicated SharedPreferences file, committed synchronously
 * so the marker survives a crash. The journal file itself is not sensitive (it
 * holds only a state enum) and is left in place by the wipe -- a leftover IDLE
 * marker is harmless. A non-IDLE marker after a reboot is exactly the signal
 * [PanicWipe.resumeIfPending] acts on.
 */
internal class FileWipeJournal(ctx: Context) : WipeJournal {
    private val prefs = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    override fun read(): PanicWipe.WipeState {
        val ord = prefs.getInt(KEY, -1)
        return if (ord < 0 || ord >= PanicWipe.WipeState.entries.size) PanicWipe.WipeState.IDLE
        else PanicWipe.WipeState.entries[ord]
    }

    override fun write(state: PanicWipe.WipeState) {
        prefs.edit().putInt(KEY, state.ordinal).commit()
    }

    override fun clear() {
        prefs.edit().clear().commit()
    }

    private companion object {
        const val PREFS = "godstone_wipe_journal"
        const val KEY = "state"
    }
}

/**
 * Production [WipeArtifacts] for Android.
 *
 * `eraseKeys` deletes the AndroidX Security AndroidKeystore master key (the KEK
 * that [Identity] and the store key prefs are encrypted under). Deleting it is
 * the cryptographic-erasure step: every EncryptedSharedPreferences ciphertext
 * becomes undecryptable regardless of whether its file is later deleted.
 * `deleteArtifacts` reuses the existing, tested store + identity panic-wipe
 * methods (DB + store-key prefs + identity prefs). `regenerateIdentity` builds a
 * fresh master key + identity via [Identity.loadOrCreate], which creates new
 * EncryptedSharedPreferences since the old ones were deleted.
 *
 * Each step is idempotent: a Keystore `deleteEntry` on an absent alias is a
 * no-op, `deleteSharedPreferences`/`deleteDatabase` on absent files are no-ops,
 * and `loadOrCreate` creates when nothing exists.
 */
internal class AndroidWipeArtifacts(private val ctx: Context) : WipeArtifacts {

    override fun eraseKeys() {
        // Crypto erasure: destroy the KEK. All EncryptedSharedPreferences
        // ciphertexts (identity, store key) become permanently undecryptable.
        val ks = KeyStore.getInstance("AndroidKeyStore")
        ks.load(null)
        if (ks.containsAlias(MASTER_KEY_ALIAS)) {
            ks.deleteEntry(MASTER_KEY_ALIAS)
        }
        // A failure to delete the master key is a genuine wipe failure; the
        // machine will re-attempt eraseKeys on resume because KEY_ERASED was
        // not persisted (run() persists only after the call returns).
    }

    override fun deleteArtifacts() {
        // Reuse the existing, tested store + identity panic-wipe methods.
        SqliteMessageStore.panicWipe(ctx)   // store DB + store key prefs
        Identity.panicWipe(ctx)             // identity prefs
    }

    override fun regenerateIdentity() {
        // Fresh master key + fresh EncryptedSharedPreferences + fresh keys.
        // The returned identity is the new node identity; nothing to do with it
        // here -- the caller (app init) will loadOrCreate again when it needs it.
        Identity.loadOrCreate(ctx)
    }

    private companion object {
        // AndroidX Security MasterKey default alias. Deleting this entry
        // invalidates every EncryptedSharedPreferences file it encrypted.
        const val MASTER_KEY_ALIAS = "_androidx_security_master_key_"
    }
}